File: apas-frontend/lib/shutdown.py
import signal
import sys
from .logger import logger

def setup_graceful_shutdown():
    def handler(signum, frame):
        logger.info(f"Received signal {signum}. Shutting down gracefully...")
        sys.exit(0)
    signal.signal(signal.SIGINT, handler)
    signal.signal(signal.SIGTERM, handler)

File: lib/shutdown.js
import logger from './logger.js';

let shutdownCallbacks = [];

export function onShutdown(callback) {
  shutdownCallbacks.push(callback);
}

export function setupGracefulShutdown() {
  const shutdown = async (signal) => {
    logger.info(`Received ${signal}. Shutting down gracefully...`);
    for (const cb of shutdownCallbacks) {
      try {
        await cb();
      } catch (err) {
        logger.error(`Shutdown callback error: ${err.message}`);
      }
    }
    process.exit(0);
  };

  process.on('SIGINT', () => shutdown('SIGINT'));
  process.on('SIGTERM', () => shutdown('SIGTERM'));
}